### ISSUE TITLE
Possible bug: vectors_to_df = FALSE still returns a dataframe

### ISSUE THREAD

**Issue by DyfanJones** [coolbutuseless/yyjsonr#13]
Not sure if this is a bug or intended behaviour. However it seems like `vectors_to_df` is ignored when set to `FALSE`.

``` r
obj <- yyjsonr::write_json_str(head(iris, 5), auto_unbox = T)
yyjsonr::read_json_str(obj, vectors_to_df = FALSE)
#>   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#> 1          5.1         3.5          1.4         0.2  setosa
#> 2          4.9         3.0          1.4         0.2  setosa
#> 3          4.7         3.2          1.3         0.2  setosa
#> 4          4.6         3.1          1.5         0.2  setosa
#> 5          5.0         3.6          1.4         0.2  setosa
```

<sup>Created on 2023-09-13 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

I was expecting the following behaviour.

``` r
obj <- yyjsonr::write_json_str(head(iris, 5), auto_unbox = T)
jsonlite::fromJSON(obj, simplifyDataFrame = F)
#> [[1]]
#> [[1]]$Sepal.Length
#> [1] 5.1
#> 
#> [[1]]$Sepal.Width
#> [1] 3.5
#> 
#> [[1]]$Petal.Length
#> [1] 1.4
#> 
#> [[1]]$Petal.Width
#> [1] 0.2
#> 
#> [[1]]$Species
#> [1] "setosa"
#> 
#> 
#> [[2]]
#> [[2]]$Sepal.Length
#> [1] 4.9
#> 
#> [[2]]$Sepal.Width
#> [1] 3
#> 
#> [[2]]$Petal.Length
#> [1] 1.4
#> 
#> [[2]]$Petal.Width
#> [1] 0.2
#> 
#> [[2]]$Species
#> [1] "setosa"
#> 
#> 
#> [[3]]
#> [[3]]$Sepal.Length
#> [1] 4.7
#> 
#> [[3]]$Sepal.Width
#> [1] 3.2
#> 
#> [[3]]$Petal.Length
#> [1] 1.3
#> 
#> [[3]]$Petal.Width
#> [1] 0.2
#> 
#> [[3]]$Species
#> [1] "setosa"
#> 
#> 
#> [[4]]
#> [[4]]$Sepal.Length
#> [1] 4.6
#> 
#> [[4]]$Sepal.Width
#> [1] 3.1
#> 
#> [[4]]$Petal.Length
#> [1] 1.5
#> 
#> [[4]]$Petal.Width
#> [1] 0.2
#> 
#> [[4]]$Species
#> [1] "setosa"
#> 
#> 
#> [[5]]
#> [[5]]$Sepal.Length
#> [1] 5
#> 
#> [[5]]$Sepal.Width
#> [1] 3.6
#> 
#> [[5]]$Petal.Length
#> [1] 1.4
#> 
#> [[5]]$Petal.Width
#> [1] 0.2
#> 
#> [[5]]$Species
#> [1] "setosa"
```

<sup>Created on 2023-09-13 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

```r
> sessioninfo::session_info()
─ Session info ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.1 (2023-06-16)
 os       macOS Ventura 13.4
 system   aarch64, darwin20
 ui       RStudio
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       Europe/London
 date     2023-09-13
 rstudio  2023.06.2+561 Mountain Hydrangea (desktop)
 pandoc   3.1.7 @ /opt/homebrew/bin/ (via rmarkdown)

─ Packages ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 package     * version date (UTC) lib source
 callr         3.7.3   2022-11-02 [1] CRAN (R 4.3.0)
 cli           3.6.1   2023-03-23 [1] CRAN (R 4.3.0)
 clipr         0.8.0   2022-02-22 [1] CRAN (R 4.3.0)
 digest        0.6.33  2023-07-07 [1] CRAN (R 4.3.0)
 evaluate      0.21    2023-05-05 [1] CRAN (R 4.3.0)
 fansi         1.0.4   2023-01-22 [1] CRAN (R 4.3.0)
 fastmap       1.1.1   2023-02-24 [1] CRAN (R 4.3.0)
 fs            1.6.3   2023-07-20 [1] CRAN (R 4.3.0)
 glue          1.6.2   2022-02-24 [1] CRAN (R 4.3.0)
 htmltools     0.5.6   2023-08-10 [1] CRAN (R 4.3.0)
 jsonlite      1.8.7   2023-06-29 [1] CRAN (R 4.3.0)
 knitr         1.43    2023-05-25 [1] CRAN (R 4.3.0)
 lifecycle     1.0.3   2022-10-07 [1] CRAN (R 4.3.0)
 magrittr      2.0.3   2022-03-30 [1] CRAN (R 4.3.0)
 pillar        1.9.0   2023-03-22 [1] CRAN (R 4.3.0)
 pkgconfig     2.0.3   2019-09-22 [1] CRAN (R 4.3.0)
 processx      3.8.2   2023-06-30 [1] CRAN (R 4.3.0)
 ps            1.7.5   2023-04-18 [1] CRAN (R 4.3.0)
 purrr         1.0.2   2023-08-10 [1] CRAN (R 4.3.0)
 R.cache       0.16.0  2022-07-21 [1] CRAN (R 4.3.0)
 R.methodsS3   1.8.2   2022-06-13 [1] CRAN (R 4.3.0)
 R.oo          1.25.0  2022-06-12 [1] CRAN (R 4.3.0)
 R.utils       2.12.2  2022-11-11 [1] CRAN (R 4.3.0)
 R6            2.5.1   2021-08-19 [1] CRAN (R 4.3.0)
 reprex        2.0.2   2022-08-17 [1] CRAN (R 4.3.0)
 rlang         1.1.1   2023-04-28 [1] CRAN (R 4.3.0)
 rmarkdown     2.24    2023-08-14 [1] CRAN (R 4.3.0)
 rstudioapi    0.15.0  2023-07-07 [1] CRAN (R 4.3.0)
 sessioninfo   1.2.2   2021-12-06 [1] CRAN (R 4.3.0)
 styler        1.10.2  2023-08-29 [1] CRAN (R 4.3.0)
 tibble        3.2.1   2023-03-20 [1] CRAN (R 4.3.0)
 utf8          1.2.3   2023-01-31 [1] CRAN (R 4.3.0)
 vctrs         0.6.3   2023-06-14 [1] CRAN (R 4.3.0)
 withr         2.5.0   2022-03-03 [1] CRAN (R 4.3.0)
 xfun          0.40    2023-08-09 [1] CRAN (R 4.3.0)
 yaml          2.3.7   2023-01-23 [1] CRAN (R 4.3.0)
 yyjsonr       0.1.9   2023-09-13 [1] Github (coolbutuseless/yyjsonr@9b99579)

 [1] /Users/dyfanjones/Library/R/arm64/4.3/library
 [2] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

**Comment by coolbutuseless** [comment]
Your `obj` produced by this line:

`obj <- yyjsonr::write_json_str(head(iris, 5), auto_unbox = T)`

in JSON this is an array of objects.

The option `vectors_to_df` controls import when the JSON is an object of equal length arrays).

``` r
library(yyjsonr)
s <- write_json_str(as.list(head(iris, 5)), auto_unbox = TRUE, pretty = TRUE)
cat(s)
#> {
#>   "Sepal.Length": [
#>     5.1,
#>     4.9,
#>     4.7,
#>     4.6,
#>     5.0
#>   ],
#>   "Sepal.Width": [
#>     3.5,
#>     3.0,
#>     3.2,
#>     3.1,
#>     3.6
#>   ],
#>   "Petal.Length": [
#>     1.4,
#>     1.4,
#>     1.3,
#>     1.5,
#>     1.4
#>   ],
#>   "Petal.Width": [
#>     0.2,
#>     0.2,
#>     0.2,
#>     0.2,
#>     0.2
#>   ],
#>   "Species": [
#>     "setosa",
#>     "setosa",
#>     "setosa",
#>     "setosa",
#>     "setosa"
#>   ]
#> }
read_json_str(s, vectors_to_df = TRUE)
#>   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#> 1          5.1         3.5          1.4         0.2  setosa
#> 2          4.9         3.0          1.4         0.2  setosa
#> 3          4.7         3.2          1.3         0.2  setosa
#> 4          4.6         3.1          1.5         0.2  setosa
#> 5          5.0         3.6          1.4         0.2  setosa
read_json_str(s, vectors_to_df = FALSE)
#> $Sepal.Length
#> [1] 5.1 4.9 4.7 4.6 5.0
#> 
#> $Sepal.Width
#> [1] 3.5 3.0 3.2 3.1 3.6
#> 
#> $Petal.Length
#> [1] 1.4 1.4 1.3 1.5 1.4
#> 
#> $Petal.Width
#> [1] 0.2 0.2 0.2 0.2 0.2
#> 
#> $Species
#> [1] "setosa" "setosa" "setosa" "setosa" "setosa"
```

<sup>Created on 2023-09-14 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by coolbutuseless** [comment]
I guess what you're after is an option to not simplify "array of object" data to a data.frame, - similar to jsonlite's `simplifyDataFrame`

Let me look. This feels possible and a worthy option to have.

**Comment by DyfanJones** [comment]
Ah yes that would be perfect :) looking forward in using this package in anger :D I am toying around the idea in adding it to [paws](https://github.com/paws-r/paws) to make it faster with it's connection to AWS 😄

**Comment by coolbutuseless** [comment]
"Fixed" this issue with b7973fc 

Summary: There are two cases where you might want to simplify to a data.frame

1. An array of objects 
2. An object of arrays

These two cases are handled separately in `yyjsonr`. By default both are set to `TRUE` meaning that `yyjsonr` will try hard to make things into a data.frame.


Changes:
* `vectors_to_df` option was renamed to `obj_of_arrs_to_df`
* Added new option: `arr_of_objs_to_df`


``` r
library(yyjsonr)
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Array of Objects
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
s <- '[{"a":1, "b":2}, {"a":2, "b": 4}]'

read_json_str(s)
#>   a b
#> 1 1 2
#> 2 2 4

read_json_str(s, arr_of_objs_to_df = FALSE)
#> [[1]]
#> [[1]]$a
#> [1] 1
#> 
#> [[1]]$b
#> [1] 2
#> 
#> 
#> [[2]]
#> [[2]]$a
#> [1] 2
#> 
#> [[2]]$b
#> [1] 4

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Object of arrays
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
s <- '{"a":[1,2], "b":[3,4]}'

read_json_str(s)
#>   a b
#> 1 1 3
#> 2 2 4

read_json_str(s, obj_of_arrs_to_df = FALSE)
#> $a
#> [1] 1 2
#> 
#> $b
#> [1] 3 4
```

<sup>Created on 2023-09-14 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by coolbutuseless** [comment]
I welcome suggestions on good names for these options :)
